### PR TITLE
Delay Compensated Asynchronous Stochastic Gradient Descent

### PR DESCRIPTION
**Pull request by just-a-jazz**
This adds the feature requested by #8744

### CODE REVIEW

**Comment by vrv** [comment]
Thanks for this change!

- There are very few comments explaining what's going on.  Because this is a variation of a standard algorithm, it probably needs some more comments explaining what the equations are.

- This is going to fail our linter because the lines are too long, indentation is not correct, etc.  See https://github.com/tensorflow/tensorflow/issues/7443#issuecomment-279182613 for what needs to be done!

- There are no tests; we need tests to make sure this is working correctly from a unit-testing perspective.

- People would want proof that this implementation can be used to validate the results of the paper -- have you done any validation yourself?  If so, can you provide some reproduction steps?  For example, having something like in the mxnet pull request linked from the issue will be useful to show: 1) that it is working, 2) that it actually helps.

Thanks again!

**Comment by vrv** [comment]
ping for @alisidd !

**Comment by alisidd** [comment]
@vrv Hi, I'm still working on it, I should hopefully have a working version by tomorrow.

**Comment by alisidd** [comment]
@vrv I've made the requested changes, and I found quite a few problems with my implementations that required a few changes that I wasn't sure if I should be making, like an addition of an optional parameter to a method in the Optimizer class. It allows for multithreaded processing using one optimizer object. Regardless, I'm sure you can see what else I had to change. 

Furthermore, I tested and validated the implementation using a modified version of mnist_softmax.py with 12 asynchronous workers, and a variance parameter of 5. I found there to be an improvement of around 1.5% in test accuracy over asynchronous stochastic gradient descent, similar to improvements found in the research paper.

**Comment by vrv** [comment]
Thanks for making these changes -- though I think changing the public API of optimizers just for this one case suggests that the Optimizer interface is now doing too much.

I'm not too familiar with the optimizer interface personally, but one thought: the idea of delay compensated updates could theoretically apply to any gradient descent algorithm; instead of having this version descend from Optimizer, does it make sense to:

1) Subclass Optimizer to be a DelayCompensatedOptimizer, where you add your specific modifications based on worker_index 

2) Make GradientDescentDC descend from it?

Let me know what you think!

(Also, I'd probably be verbose on the names here: DelayCompensatedGradientDescent; the term "DC" isn't standard enough for it to be abbreviated).

**Comment by vrv** [comment]
Also there aren't any tests that actually exercise more than 1 worker, so it's hard to tell the context in which it's needed.

@jhseu in case he has any thoughts on optimizer interface

**Comment by alisidd** [comment]
@vrv A whole class for adding the ability to address updates by the worker index would have usage if more algorithms were added similar to this one that depended on having that kind of information. I need that information because I need to store a separate var for each worker, and keep them separate throughout the optimization. I don't think most gradient descent algorithms need that kind of information.

**Comment by vrv** [comment]
(I've assigned this the API review label so that the folks who look at API changes can assess whether this is something to accept.)

**Comment by alisidd** [comment]
@vrv I can also copy all relevant code to my new file and override the relevant methods if this is too drastic a change to the Optimizer class.

**Comment by vrv** [comment]
It would certainly avoid an API review if you did this.  In addition, it might also make sense for this to be in contrib/opt like some other optimizers, so that we can evolve the delay compensated optimizers further.  All code in core has API stability guarantees that will far outlast the initial maintainers :)

**Comment by asimshankar** [comment]
In terms of API review, +1 to the last two suggestions:

- Put this optimizer in contrib
- Avoid changing the Optimizer interface and override the relevant methods in the subclass

**Comment by alisidd** [comment]
@vrv I've made the required changes.

File: tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py
Comment (vrv) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: 2017

File: tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py
Comment (vrv) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: two whitespace lines between every top line function

File: tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py
Comment (vrv) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: two space indent in all python files

see https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md#python-coding-style in general

File: tensorflow/python/training/optimizer.py
Comment (vrv) [inline tensorflow/python/training/optimizer.py]: are the changes in this file still required?

File: tensorflow/core/ops/training_ops.cc
Comment (vrv) [inline tensorflow/core/ops/training_ops.cc]: Sorry for not seeing this before: can you update this to use the "resource type" like in ResourceApplyGradientDescent instead?  We're moving to "Resource variables" which have meaningful type semantics (refs don't).

File: tensorflow/core/ops/training_ops.cc
Comment (vrv) [inline tensorflow/core/ops/training_ops.cc]: Is there a better name that 'var_bak' we can use?  maybe like 'shadow' or something?

File: tensorflow/core/ops/training_ops.cc
Comment (vrv) [inline tensorflow/core/ops/training_ops.cc]: parameter

**Review by vrv — CHANGES_REQUESTED** [review]

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
@alisidd sorry it looks like you need to rebase again.

**Comment by alisidd** [comment]
@rmlarsen Should be good now

File: tensorflow/core/ops/training_ops.cc
Comment (alisidd) [inline tensorflow/core/ops/training_ops.cc]: Did you also want me to rename the op to "ResourceApplyDelayCompensatedGradientDescent"?

**Comment by vrv** [comment]
No need to change the name!  @tensorflow-jenkins test this please

**Comment by alisidd** [comment]
#10055 seems to have had the same problem when it was approved, not sure if I'm supposed to do something.

**Review by vrv — APPROVED** [review]

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

(it looks like no GPU kernels are registered, but perhaps that can be added as a future PR if someone wants.  it's pretty straightforward to copy the existing code for other optimizers).

**Comment by fangy** [comment]
@alisidd  would you like share your modified version of mnist_softmax.py ? I wanna have a try. thx .

**Comment by zhuhyc** [comment]
@alisidd @vrv 
I add supporting GPU kernels base your code, when I execute a test,it runs on CPU normally, but it occurs an error  on GPU

     python tensorflow/tensorflow/examples/tutorials/mnist/mnist_softmax_dc_asgd.py

the error is:

      **_2017-06-28 10:58:29.296340: E tensorflow/core/common_runtime/executor.cc:641] Executor failed to create kernel. Invalid argument: AttrValue must not have reference type value of float_ref
	 for attr 'tensor_type'
	; NodeDef: w/DelayCompensatedGradientDescent/_23 = _HostRecv[_start_time=0, client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_103_w/DelayCompensatedGradientDescent", tensor_type=DT_FLOAT_REF, _device="/job:localhost/replica:0/task:0/gpu:0"](^w/DelayCompensatedGradientDescent/_22, ^DelayCompensatedGradientDescent/learning_rate, ^gradients/MatMul_grad/tuple/control_dependency_1, ^DelayCompensatedGradientDescent/lambda); Op<name=_HostRecv; signature= -> tensor:tensor_type; attr=tensor_type:type; attr=tensor_name:string; attr=send_device:string; attr=send_device_incarnation:int; attr=recv_device:string; attr=client_terminated:bool,default=false; is_stateful=true>
	 [[Node: w/DelayCompensatedGradientDescent/_23 = _HostRecv[_start_time=0, client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_103_w/DelayCompensatedGradientDescent", tensor_type=DT_FLOAT_REF, _device="/job:localhost/replica:0/task:0/gpu:0"](^w/DelayCompensatedGradientDescent/_22, ^DelayCompensatedGradientDescent/learning_rate, ^gradients/MatMul_grad/tuple/control_dependency_1, ^DelayCompensatedGradientDescent/lambda)]]_**


the test code is here : 
     https://github.com/zhuhyc/tensorflow/blob/dc-asgd/tensorflow/examples/tutorials/mnist/mnist_softmax_dc_asgd.py

the patch is here: 
     https://github.com/zhuhyc/tensorflow/blob/dc-asgd/dc-asgd.patch
can you give me some suggestions to reslove this error?

**Comment by sesamesub** [comment]
I'm trying to learn more dcsgd and external optimizers for tf, I'm unsure how to implement this feature with the tf learn library examples. 

@zhuhyc 
Does your testcode work for CPUs? When I run the softmax example on docker nightly CPU build, I get the following result:

 $ python mnist_softmax_dc_asgd.py

> Traceback (most recent call last):
  File "mnist_softmax_dc_asgd.py", line 84, in <module>
    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/platform/app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "mnist_softmax_dc_asgd.py", line 64, in main
    train_step = optimizer.minimize(cross_entropy,worker_index=0)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py", line 144, in minimize
    name=name, worker_index=worker_index)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py", line 217, in apply_gradients
    worker_index))
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py", line 34, in update_op_asynchronous
    return optimizer._apply_dense(g, self._v, index)
TypeError: _apply_dense() takes exactly 3 arguments (4 given)

>

**Comment by zhuhyc** [comment]
@sesamesub  
    Yes, it works for on CPUs correctly. My tf version is master branch and have applyed dc-asgd.patch
    the patch is here:
        https://github.com/zhuhyc/tensorflow/blob/dc-asgd/dc-asgd.patch
    It looks like you don't apply the patch, please apply it and test again

**Comment by fangy** [comment]
ping to check status

**Comment by sesamesub** [comment]
I had some issues on my machine which I fixed, your code/patch is fine for
CPU tensorflow and works for per your example code.

**Comment by zhuhyc** [comment]
Ping to check status @alisidd

File: tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py
Comment (ry) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: Hello -

Comment (ry) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: Why not use the `minimize` from the parent class? It seems this is copied exactly?

File: tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py
Comment (ry) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: Also seems to be copied from parent class?

Comment (ry) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: @alisidd question for you above regarding subclassing. 
cc @vrv

Comment (vrv) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: My preference was that delay-compensated-specific bits should not leak out to the parent class, which the original implementation did by adding worker_index to the public APIs.  Where possible, it would be indeed nice to just reuse existing funtionality like in minimize(), but this function relies on extra state that is not available in the base APIs.

**Comment by MattX** [comment]
It looks like this PR was removed from master (commit 47949be1aae1d7c97f75712e0655da96fab90208), with the message "Remove Delay Compensated Asynchronous Stochastic Gradient Descent (DCASGD). The implementation is broken in its current form."

What was wrong with the implementation? The message only references this PR and issue, none of which seem to be raising issues.

**Comment by vrv** [comment]
IIRC, it was lacking an _apply_dense() implementation, there were typos that caused failures when used, etc.

The commit is obviously in the history for somebody to re-use and fix, if they wanted, but they'd need to write more tests and probably refactor it to make it something that the team could accept to TensorFlow.

**Pull request merged**
